File: rhino-update.rs
// rhino-update – colourful one-shot update & cleanup for Rhino Linux
// Requires sudo.

use std::env;
use std::io::{self, Write};
use std::process::{Command, Stdio};

// ANSI helpers ----------------------------------------------------------------
const RESET: &str = "\x1b[0m";
const BOLD: &str = "\x1b[1m";
const RED: &str = "\x1b[31m";
const GREEN: &str = "\x1b[32m";
const YELLOW: &str = "\x1b[33m";
const BLUE: &str = "\x1b[34m";
const MAGENTA: &str = "\x1b[35m";
const CYAN: &str = "\x1b[36m";

macro_rules! color_print {
    ($color:expr, $($arg:tt)*) => {{
        print!("{}{}", $color, format_args!($($arg)*));
        print!("{}", RESET);
    }};
}

// Run a command, streaming its output, and exit on failure --------------------
fn run(cmd: &[&str], description: &str) -> io::Result<()> {
    if !description.is_empty() {
        color_print!(format!("{BLUE}{BOLD}➜ {RESET}{}"), description);
        println!();
    }
    color_print!("{}▶ ", CYAN);
    println!("{}", cmd.join(" "));

    let mut iter = cmd.iter();
    let program = iter.next().unwrap_or(&"");
    let status = Command::new(program)
        .args(iter)
        .stdin(Stdio::null())
        .stdout(Stdio::inherit())
        .stderr(Stdio::inherit())
        .status()?;

    if !status.success() {
        color_print!("{RED}❌ Command failed with exit code: {:?}\n", status.code());
        std::process::exit(status.code().unwrap_or(1));
    }
    Ok(())
}

fn main() {
    if env::uid() != 0 {
        color_print!("{RED}❌ This script must be run as root (sudo).\n");
        std::process::exit(1);
    }

    color_print!(
        "{MAGENTA}{BOLD}🦏 Rhino Linux Update & Cleanup{}\n\n",
        RESET
    );

    let result = || -> io::Result<()> {
        run(&["rpk", "update", "-y"], "Updating all packages …")?;
        run(&["rpk", "cleanup", "-y"], "Purging orphaned packages …")?;
        Ok(())
    }();

    if let Err(e) = result {
        color_print!("{RED}❌ Error: {}\n", e);
        std::process::exit(1);
    }

    color_print!("{GREEN}✅ Rhino Linux is up-to-date and squeaky-clean!\n");
}

File: sysadmin/rhino-update.py
#!/usr/bin/env python3
"""
rhino-update – colourful one-shot update & cleanup for Rhino Linux
Requires sudo.
"""

import os
import subprocess
import sys
from typing import List

# ── ANSI helpers ----------------------------------------------------------- #
RESET   = "\033[0m"
BOLD    = "\033[1m"
RED     = "\033[31m"
GREEN   = "\033[32m"
YELLOW  = "\033[33m"
BLUE    = "\033[34m"
MAGENTA = "\033[35m"
CYAN    = "\033[36m"

def ok(msg: str) -> None:
    print(f"{GREEN}✅ {msg}{RESET}")

def warn(msg: str) -> None:
    print(f"{YELLOW}⚠️  {msg}{RESET}")

def err(msg: str) -> None:
    print(f"{RED}❌ {msg}{RESET}")

def info(msg: str) -> None:
    print(f"{CYAN}ℹ️  {msg}{RESET}")

def title(msg: str) -> None:
    print(f"\n{MAGENTA}{BOLD}🦏 {msg}{RESET}")

# ── Core logic ------------------------------------------------------------- #
def run(cmd: List[str], description: str = "") -> None:
    if description:
        print(f"\n{BLUE}{BOLD}➜ {description}{RESET}")
    print(f"{CYAN}▶ {' '.join(cmd)}{RESET}")

    try:
        subprocess.run(cmd, check=True, stdout=sys.stdout, stderr=sys.stderr)
    except subprocess.CalledProcessError as exc:
        err(f"Command failed (exit code {exc.returncode})")
        print(f"{RED}Failed: {' '.join(exc.cmd)}{RESET}", file=sys.stderr)
        sys.exit(exc.returncode)
    except FileNotFoundError as exc:
        err(f"Command not found: {exc.filename}")
        sys.exit(127)

def main() -> None:
    if os.geteuid() != 0:
        err("This script must be run as root (sudo).")
        sys.exit(1)

    title("Rhino Linux Update & Cleanup")

    try:
        run(["rpk", "update", "-y"], "Updating all packages …")
        run(["rpk", "cleanup", "-y"], "Purging orphaned packages …")

    except KeyboardInterrupt:
        warn("Operation cancelled by user.")
        sys.exit(130)

    ok("Rhino Linux is up-to-date and squeaky-clean!")

if __name__ == "__main__":
    main()
